File: Strand/BLE/StandardHeartRate.swift
import Foundation
import WhoopProtocol

/// Pure parser for the standard BLE Heart Rate Measurement characteristic (0x2A37).
/// Returns the heart rate (bpm), R-R intervals (ms), raw R-R ticks (u16), and the contact state.
///
/// `rr` contains milliseconds converted from 1/1024-second units per the BLE spec.
/// `rrRawTicks` contains the raw u16 values before conversion. WHOOP 5 (firmware 50.41.1.0)
/// sends millisecond values on this characteristic, non-compliant with the BLE spec — callers
/// that know they have a WHOOP 5 should use `rrRawTicks` instead of `rr`.
public enum StandardHeartRate {
    public static func parse(_ data: [UInt8]) -> (hr: Int, rr: [Int], rrRawTicks: [Int], contact: StandardHRContact)? {
        guard !data.isEmpty else { return nil }
        let flags = data[0]
        let contact = StandardHRContact.fromMeasurementFlags(flags)
        var idx = 1
        let hr: Int
        if flags & 0x01 != 0 {                       // 16-bit HR
            guard idx + 1 < data.count else { return nil }
            hr = Int(data[idx]) | (Int(data[idx + 1]) << 8); idx += 2
        } else {                                     // 8-bit HR
            guard idx < data.count else { return nil }
            hr = Int(data[idx]); idx += 1
        }
        if flags & 0x08 != 0 { idx += 2 }            // skip Energy Expended (bit 3)
        var rr: [Int] = []
        var rrRawTicks: [Int] = []
        if (flags >> 4) & 0x01 != 0 {                // R-R present (bit 4)
            while idx + 1 < data.count {
                let raw = Int(data[idx]) | (Int(data[idx + 1]) << 8)
                rrRawTicks.append(raw)
                rr.append(Int((Double(raw) / 1024.0 * 1000.0).rounded()))   // 1/1024 s → ms
                idx += 2
            }
        }
        return (hr, rr, rrRawTicks, contact)
    }
}

File: android/app/src/main/java/com/noop/ble/StandardHeartRate.kt
package com.noop.ble

import com.noop.protocol.StandardHrContact

/**
 * Pure parser for the standard BLE Heart Rate Measurement characteristic (0x2A37).
 *
 * Faithful Kotlin twin of Strand/BLE/StandardHeartRate.swift. Returns the heart rate (bpm) and any
 * R-R intervals (ms). Pure → unit-testable away from android.bluetooth.
 *
 * This is a SEPARATE parser from [WhoopBleClient.parseStandardHr] on purpose: the new isolated
 * [StandardHrSource] uses THIS one so the WHOOP client's inline parse stays untouched (slight
 * duplication is fine — it keeps the hardware-verified WHOOP path from regressing). Both encode the
 * same Bluetooth SIG layout:
 *   - flags bit0 (0x01): HR is u16 (else u8)
 *   - flags bit3 (0x08): Energy-Expended field present → skip its 2 bytes before R-R
 *   - flags bit4 (0x10): one or more R-R intervals follow, each a u16 in 1/1024-second units
 *
 * R-R is converted to milliseconds as `round(raw / 1024 * 1000)` to match the Swift parser exactly
 * (the WHOOP store keeps R-R in ms).
 */
object StandardHeartRate {

    /** The parsed reading: heart rate (bpm) and the R-R intervals (ms), in arrival order. */
    data class Reading(val hr: Int, val rr: List<Int>, val contact: StandardHrContact)

    /**
     * Parse one 0x2A37 notification payload. Returns null on an empty or truncated packet (a packet
     * whose declared HR/R-R bytes run past the buffer), matching the Swift `guard` bounds checks.
     */
    fun parse(data: ByteArray): Reading? {
        if (data.isEmpty()) return null
        val flags = data[0].toInt() and 0xFF
        val contact = StandardHrContact.fromMeasurementFlags(flags)
        var idx = 1

        val hr: Int
        if (flags and 0x01 != 0) {                       // 16-bit HR
            if (idx + 1 >= data.size) return null
            hr = (data[idx].toInt() and 0xFF) or ((data[idx + 1].toInt() and 0xFF) shl 8)
            idx += 2
        } else {                                         // 8-bit HR
            if (idx >= data.size) return null
            hr = data[idx].toInt() and 0xFF
            idx += 1
        }

        if (flags and 0x08 != 0) idx += 2                // skip Energy Expended (bit 3)

        val rr = ArrayList<Int>()
        if ((flags shr 4) and 0x01 != 0) {               // R-R present (bit 4)
            while (idx + 1 < data.size) {
                val raw = (data[idx].toInt() and 0xFF) or ((data[idx + 1].toInt() and 0xFF) shl 8)
                rr.add(Math.round(raw / 1024.0 * 1000.0).toInt())   // 1/1024 s → ms (rounded)
                idx += 2
            }
        }
        return Reading(hr, rr, contact)
    }
}
